### PR TITLE
feat(eap): Make EAP RPC endpoint use query cache

### PR DESCRIPTION
**Pull request by colin-sentry**
This returns makes the RPC graph endpoint use the -State for aggregates and instructs clickhouse to cache the results.

Before this change, a visit to RPC graph would result in this clickhouse query:
```
SELECT quantileTDigestWeighted(weight)(0.5)(blah), time
FROM spans
WHERE _sort_timestamp >= (last week) and _sort_timestamp < (right now)
GROUP BY toStartOfDay(_sort_timestamp) AS time
ORDER BY time
```

After this change, you get 24 separate queries that look like this:
```
SELECT  quantileTDigestWeightedState(weight)(0.5)(blah)
FROM spans
WHERE _sort_timestamp >= (last week rounded to start of day)
AND _sort_timestamp < (6 days ago rounded to start of day)
SETTINGS use_query_cache=true, query_cache_ttl=7776000
```

The use_query_cache setting means clickhouse can immediately return if any hour-long chunk has ever been used in another query.
You could even expose 2-hour intervals in the frontend, and this would still make 1 hour-sized requests to clickhouse, then merge them before returning from the RPC.

### CODE REVIEW

**Review by evanh — COMMENTED** [review]
I think you would need to account for the increase in concurrent queries that this type of splitting will cause. We'd probably need to increase the limits in our own allocation policies or Clickhouse or both.

These individual queries will also get cached in Redis, so do we actually need `use_query_cache`?

I also think we will want some more logic in here to ensure we are refreshing newer data more frequently than older data. As in, skip the query cache for the most recent four hour chunk or something like that.

**Review by evanh — COMMENTED** [review]
Also, can you write a test for this functionality?

**Comment by colin-sentry** [comment]
> As in, skip the query cache for the most recent four hour chunk or something like that.
done and tested

> Also, can you write a test for this functionality?
done

File: snuba/web/rpc/timeseries/aggregate_functions.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+# https://github.com/ClickHouse/ClickHouse/blob/f7ca33868b976b92499178d475a21fd72e9badfa/src/AggregateFunctions/QuantileTDigest.h#L104
+def interpolate(x: float, x1: float, y1: float, x2: float, y2: float) -> float:
+    k = (x - x1) / (x2 - x1)
+    return (1 - k) * y1 + k * y2
+
+
+# this is a port of quantileTDigestMerge(...) in clickhouse, in this file:
+# https://github.com/ClickHouse/ClickHouse/blob/f7ca33868b976b92499178d475a21fd72e9badfa/src/AggregateFunctions/QuantileTDigest.h
+def merge_t_digests_states(states: Iterable[str], level: float) -> float:
```
[the remark this answers is not in the document]
Comment (colin-sentry) [reply snuba/web/rpc/timeseries/aggregate_functions.py:38]: there's only one, and it requires a dict allocation per centroid, which would be non-negligible for us (1000+ centroids per query)

[the remark this answers is not in the document]
Comment (colin-sentry) [reply snuba/web/rpc/timeseries/aggregate_functions.py:38]: also it does a bunch of other stuff like compressing, I think it's fine to use this code, it's just direct port from clickhouse anyways

[the remark this answers is not in the document]
Comment (colin-sentry) [reply snuba/web/rpc/timeseries/aggregate_functions.py:38]: if you want to use a library, we could just add numpy to snuba, though that again seems like overkill? we only need the "weighted percentile" function https://numpy.org/doc/stable/reference/generated/numpy.percentile.html

**Review by phacops — APPROVED** [review]

**Comment by colin-sentry** [comment]
> These individual queries will also get cached in Redis, so do we actually need `use_query_cache`?

Just to follow up - the plan is to just bypass the entire snuba query pipeline for now and re-add the chunks we need. The clickhouse readthrough cache should have less overhead than the redis one for the large number of queries this endpoint ends up making

> I also think we will want some more logic in here to ensure we are refreshing newer data more frequently than older data. As in, skip the query cache for the most recent four hour chunk or something like that.

This is done!

> I think you would need to account for the increase in concurrent queries that this type of splitting will cause. We'd probably need to increase the limits in our own allocation policies or Clickhouse or both.

Yes, there is similar issue with alerts, we should discuss how to do rate limiting for the RPCs and what allocation policies we want to use, esp. if we want to support 1bn+ span queries for some customers

**Comment by evanh** [comment]
> the plan is to just bypass the entire snuba query pipeline for now and re-add the chunks we need.

What chunks do you think you won't need?

> The clickhouse readthrough cache should have less overhead than the redis one

The purpose of the Redis cache is to keep queries off of Clickhouse. That prevents an unnecessary network call, and also avoids us having to factor in cached queries to our allocation policies. I think we should use Redis for this cache for sure.

File: snuba/web/rpc/timeseries/aggregate_functions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (evanh) [reply snuba/web/rpc/timeseries/aggregate_functions.py:38]: If there's no library to use then can you add some unit tests for this please.

**Pull request merged**